File: html2myst.js

var contents = '';

var mainSection = document.querySelector('body > div.section');

var className = document.querySelector('body > div.section > div.titlepage').innerText;

contents += `# ${className}\n\n`;

var sections = document.querySelectorAll('body > div.section > div.section');

for (var ix = 0; ix < sections.length; ++ix)
{
	var section = sections[ix];

	var sectionName = [...section.children].filter(el => el.matches('div.titlepage'));

	contents += `## ${sectionName[0].innerText}\n\n`;

	var subSections = [...section.children].filter(el => el.matches('div.section'));
	
	for (var iy = 0; iy < subSections.length; ++iy)
	{
		contents += '::::{abi-group}\n';

		var subSection = subSections[iy];

		for (var iz = 0; iz < subSection.children.length; ++iz)
		{
			var child = subSection.children[iz];

			if (child.matches('div.titlepage'))
				continue;
			
			if (child.matches('code')) {
				dumpDeclaration(child);
				continue;
			}

			if (child.matches('div.synopsisgroup')) {
				for (var iw = 0; iw < child.children.length; ++iw) {
					if (child.children[iw].matches('code')) {
						dumpDeclaration(child.children[iw]);
					}
				}
			}

			contents += '\n';
			dumpNode(child);
		}		

		contents += '::::\n\n';
	}
}

function dumpDeclaration(code)
{
	var text = code.innerText
		.replaceAll('\n', ' ')
		.replaceAll(/[\t \xa0]+/g, ' ') // tab, space, nbsp
		.replace(/;$/, '');
	if (text.startsWith('global') && text.indexOf('operator') >= 0) {
		// global operator overloading... return as-is
		contents += `:::{cpp:function} ${text.replace(/^global/, '')}\n:::\n`;
	} else {
		let isGlobal = text.endsWith('global');
		if (isGlobal) {
			// non-member functions, in the global namespace
			contents += `:::{cpp:function} ${text.replace(/global$/, '')}\n:::\n`;
		} else {
			text = text.replace(/([a-zA-Z0-9_]+)\(/,
					(match, name, offset, string, groups) => {
						return `${className}::${name}(`;
					})
					.replace(/(operator[+\-*\/%^&|~!=<>!,()[\]]{1,3})\(/,
					(match, name, offset, string, groups) => {
						return `${className}::${name}(`;
					});
			contents += `:::{cpp:function} ${text}\n:::\n`;
		}
	}
}

function dumpNode(element)
{
	if (element.matches('p'))
	{
		dumpParagraph(element);
	}
	else if (element.matches('div') && element.className.indexOf('admonition') >= 0)
	{
		let paragraphs = element.querySelectorAll('p');
		let title = element.querySelector('div.title');
		let type = element.className.replace('admonition', '').trim();

		contents += `:::{admonition} ${title.innerText}\n`;
		contents += `:class: ${type}\n`;

		for (let ix = 0; ix < paragraphs.length; ++ix) {
			dumpParagraph(paragraphs[ix]);
		}

		contents += '\n:::\n\n';
	}
	else if (element.matches('pre'))
	{
		contents += ':::{code}\n';
		contents += element.textContent;
		contents += '\n:::\n';
	}

	if (element.matches('table'))
	{
		// output a table... should be able to automate this too...
		contents += ':::{list-table}\n:::\n\n';
	}
}

function dumpParagraph(paragraph)
{
	for (var ix = 0; ix < paragraph.childNodes.length; ++ix)
	{
		var node = paragraph.childNodes[ix];
		
		if (node.nodeName == '#text')
		{
			contents += node.textContent;
		}
		else
		{
			dumpInline(node);
		}
	}

	contents += '\n\n';
}

function dumpInline(node)
{
	if (node.matches('code'))
	{
		var highlight = '';
		switch (node.className)
		{
			case 'methodname':	highlight = 'hmethod'; break;
			case 'classname':	highlight = 'hclass'; break;
			case 'parameter':	highlight = 'hparam'; break;
			default: break;
		}

		if (highlight.length > 0)
			contents += `{${highlight}}\`${node.textContent}\``;
		else
			contents += node.textContent;
	}
	else if (node.matches('a'))
	{
		var codeNode = node.firstChild;
		var type = '';
		switch (codeNode.className)
		{
			case 'methodname':	type = 'cpp:func'; break;
			case 'classname':	type = 'cpp:class'; break;
			default:			type = '...'; break;
		}

		var result = node.attributes['href'].value.match(/#([^_]+)_([^_]+)/);

		if (result != null)
		{
			let className = result[1];
			let methodName = result[2];

			if (codeNode.textContent.indexOf('::') < 0)
				contents += `{cpp:func}\`~${className}::${methodName}\``;
			else
				contents += `{cpp:func}\`${className}::${methodName}\``;
		}
		else
		{
			contents += `{${type}}\`${codeNode.textContent}\``;
		}
	}
	else if (node.matches('span'))
	{
		if (node.className == 'code')
			contents += `\`${node.textContent}\``;
		else
			contents += node.textContent;
	}
	else
	{
		contents += node.textContent;
	}
}

var lines = contents.split('\n');
for (let ix = 0; ix < lines.length; ++ix)
{
	lines[ix] = lines[ix].trimEnd();
}

contents = lines.join('\n')
	.replaceAll(/[\n]+\n/g, '\n\n');

console.log(contents);

File: html2myst.py
#!/usr/bin/env python3

from bs4 import BeautifulSoup
from bs4.builder import XMLParsedAsHTMLWarning
import warnings
import re
import sys

combine_ws = re.compile(r'\s+')
function_re = re.compile(r'([a-zA-Z0-9_~]+)\(')
operator_re = re.compile(r'(operator[+\-*\/%^&|~!=<>,()[\]]{1,3})\(')
reference_re = re.compile(r'#([^_]+)_([^_]+)')

warnings.filterwarnings('ignore', category=XMLParsedAsHTMLWarning)

className = ''

def text(elem, clean=True):
	if clean:
		return combine_ws.sub(' ', ''.join(elem.strings)).strip()
	else:
		return ''.join(elem.strings).strip()

def find_class(elem, tagname, cls):
	elem.find(tagname, class_=cls, recursive=False)

def findall_class(elem, name, cls):
	elem.find_all(name, class_=cls, recursive=False)

def has_class(elem, name):
	return elem.has_attr('class') and name in elem['class']

def add_class_name(match):
	return f'{className}::{match.group(1)}('


nodeNames = []

abigroups = []

def process_link(anchor):
	match = reference_re.search(anchor['href'])
	refType = 'ref'

	code = anchor.select_one(':scope > code')
	if code is not None:
		className = code['class'][0]
		if className == 'methodname':
			refType = 'cpp:func'
		elif className == 'classname':
			refType = 'cpp:class'
			
	textContent = text(anchor)

	if match:
		refClass = match.group(1)
		refMethod = match.group(2)
		if '::' in textContent:
			return f'{{cpp:func}}`{refClass}::{refMethod}`'
		else:
			return f'{{cpp:func}}`~{refClass}::{refMethod}`'
	else:
		return f'{{{refType}}}`{textContent}`'

def process_function(code):
	function = text(code)

	if function.startswith('global') and 'operator' in function:
		function = function[len('global'):].strip()
	elif function.endswith('global'):
		function = function[0:-len('global')]
	else:
		function = function_re.sub(add_class_name, function)
		function = operator_re.sub(add_class_name, function)
	
	function = function.replace(';', '')

	return f':::{{cpp:function}} {function}\n:::'

def process_content(node):
	if not hasattr(node, 'contents'):
		# node is text content, return as-is
		return node.string
	
	content = ''
	for child in node.children:
		if not hasattr(child, 'contents'):
			content += child.string
		elif child.name == 'p':
			content += process_content(child).strip().replace('\n', ' ')
		elif child.name == 'code':
			highlight = None
			is_enum = False
			className = '' if not hasattr(child, 'class') else child['class'][0]
			if className == 'methodname':
				highlight = 'hmethod'
			elif className == 'classname':
				highlight = 'hclass'
			elif className == 'parameter':
				highlight = 'hparam'
			elif className == 'constant':
				is_enum = True
			
			if is_enum:
				content += f'{{cpp:enum}}`{text(child)}`'
			elif highlight != None:
				content += f'{{{highlight}}}`{text(child)}`'
			else:
				content += text(child)
		elif child.name == 'a':
			content += process_link(child)
		elif child.name == 'span':
			if 'code' in child['class']:
				content += f'`{text(child)}`'
			elif 'term' in child['class']:
				for grandchild in child.children:
					if not hasattr(grandchild, 'contents'):
						content += text(grandchild)
					elif grandchild.name == 'code' and 'constant' in grandchild['class']:
						content += f'{{cpp:enum}}`{text(grandchild)}`'
					else:
						content += text(grandchild)
			elif 'type' in child['class']:
				content += f'{{htype}}`{text(child)}`'
			elif 'keycap' in child['class']:
				content += f'{{hkey}}`{text(child)}`'
				print('keycap', text(child))
			else:
				print('WARNING: unable to handle span with classes:', child['class'])
				print('    ', text(child))
				content += text(child)
		elif child.name == 'div' or child.name == 'table':
			content += '\n'.join(process_block(child))
		else:
			print('WARNING: unable to handle child of type:', child.name)
			content += text(child)
	
	return content

def process_block(node):
	lines = []
	if (node.name == 'p'):
		lines.append(process_content(node))
	elif node.name == 'div' and has_class(node, 'admonition'):
		title = text(node.select_one(':scope > div.title'))
		kind = ''.join([cls for cls in node['class'] if cls != 'admonition'])
		lines.append(f':::{{admonition}} {title}')
		lines.append(f':class: {kind}')

		for child in node.select(':scope p'):
			lines.append(process_content(child))
		
		lines.append(':::')
	elif node.name == 'pre':
		lines.append(':::{code}')
		lines.append(text(node, clean=False))
		lines.append(':::')
	elif node.name == 'table':
		lines.append(':::{list-table}')
		lines.append('---')
		lines.append('header-rows: 1')
		lines.append('align: left')
		lines.append('widths: auto')
		lines.append('---')
		lines.append('')
		rows = node.select('tr')
		for row in rows:
			lines.append('-')
			lines.append('')
			for column in row.select('th, td'):
				lines.append(f'\t- {process_content(column)}')
				lines.append('')
		lines.append(':::')
	elif node.name == 'ul':
		for item in node.select('li'):
			lines.append(f'- {process_content(item)}')
			lines.append('')
	else:
		lines.append(process_content(node))
	
	return lines

def main():
	global className
	lines = []

	with open(sys.argv[1]) as file:
		soup = BeautifulSoup(file, 'lxml')

	print(f'Processing {sys.argv[1]}...')

	mainSection = soup.select_one('body > div.section')

	className += text(mainSection.select_one(':scope > div.titlepage'))
	lines.append(f':::{{cpp:class}} {className}')
	lines.append(':::')
	lines.append('')
	#print(className)
	lines.append(f'# {className}')

	sections = mainSection.select(':scope > div.section')

	for section in sections:
		sectionName = text(section.select_one(':scope > div.titlepage'))
		#print('  ', sectionName)
		lines.append(f'## {sectionName}')

		for subSection in section.select(':scope > div.section'):
			lines.append('::::{abi-group}\n\n')
			for node in subSection.children:
				# just to figure out node types we're seeing...
				if node.name not in nodeNames:
					nodeNames.append(node.name)
				
				# known children...
				# - div
				#	- admonition (:::{admonition})
				#	- synopsisgroup (group of :::{cpp:function})
				# - p: formatted paragraph content
				# - ul: formatted list
				# - table: (:::{list-table})
				# - code: (:::{cpp:function})
				# - pre: (:::{code})
				if node.name == 'div' and has_class(node, 'titlepage'):
					continue
				elif node.name == 'code':
					#print('	', process_function(node))
					lines.append(process_function(node))
					lines.append('')
				elif node.name == 'div' and has_class(node, 'synopsisgroup'):
					for code in node.select(':scope > code'):
						#print('	', process_function(code))
						lines.append(process_function(code))
					lines.append('')
				else:
					lines.extend(process_block(node))
					lines.append('')
		
			lines.append('::::\n\n')			
			
	#print('\n'.join(lines).replace('\n\n', '\n'))
	with open(sys.argv[2], 'w') as file:
		file.write('\n'.join(lines).replace('\n\n', '\n'))

if __name__ == "__main__":
	main()


